allocator/allocator.go

```diff
@@ -0,0 +1,91 @@
+package allocator
+
+import (
+	"strconv"
+
+	"github.com/ipfs/ipfs-cluster/api"
+
+	rpc "github.com/hsanjuan/go-libp2p-gorpc"
+	cid "github.com/ipfs/go-cid"
+	peer "github.com/libp2p/go-libp2p-peer"
+)
+
+type Allocator interface {
+	SetClient(c *rpc.Client)
+	Shutdown() error
+	// Allocate returns where to allocate a pin request based on metrics which
+	// carry a numeric value such as "used disk". We do not pay attention to
+	// the metrics of the currently allocated peers and we just sort the
+	// candidates based on their metric values.
+	Allocate(c *cid.Cid, current, candidates map[peer.ID]api.Metric) ([]peer.ID, error)
+}
+
+// SimpleAllocator implements the Allocator interface. Other allocators can
+// extend SimpleAllocator and should at least override the Allocate() function.
+type SimpleAllocator struct{}
+
+// SetClient does nothing in this allocator
+func (alloc SimpleAllocator) SetClient(c *rpc.Client) {}
+
+// Shutdown does nothing in this allocator
+func (alloc SimpleAllocator) Shutdown() error { return nil }
+
+// Allocate does nothing in this allocator (important to override this)
+func Allocate(c *cid.Cid, current, candidates map[peer.ID]api.Metric) ([]peer.ID, error) {
+	return nil, nil
+}
+
+// MetricsSorter attaches sort.Interface methods to our metrics and sorts
+// a slice of peers in the way that interest us
+type MetricsSorter interface {
+	Len() int
+	Swap(i, j int)
+	Less(i, j int) bool
+}
+
+// SimpleSorter implements the MetricsSorter interface. Other sorters may extend
+// this and should at least override the default Less() function.
+type SimpleSorter struct {
```

Comment:
Probably remove the MetricSorter interface and rename this to MetricSorter (a type that implements the sort.Interface for metrics). The type can have a property `Reverse` which would allow it to sort in reverse order.